src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-cluster-form/smb-cluster-form.component.ts

```diff
@@ -0,0 +1,258 @@
+import { ModalCdsService } from '~/app/shared/services/modal-cds.service';
+import { HostService } from '~/app/shared/api/host.service';
+import { SmbService } from '~/app/shared/api/smb.service';
+import { TaskWrapperService } from '~/app/shared/services/task-wrapper.service';
+import { SmbDomainSettingModalComponent } from '../smb-domain-setting-modal/smb-domain-setting-modal.component';
+
+@Component({
+  selector: 'cd-smb-cluster-form',
+  templateUrl: './smb-cluster-form.component.html',
+  styleUrls: ['./smb-cluster-form.component.scss']
+})
+export class SmbClusterFormComponent extends CdForm implements OnInit {
+  smbForm: CdFormGroup;
+  hostsAndLabels$: Observable<{ hosts: any[]; labels: any[] }>;
+  hasOrchestrator: boolean;
+  orchStatus$: Observable<any>;
+  publicAddrPattern = /^(\d{1,3}\.){3}\d{1,3}\/\d{1,2}(\%[a-zA-Z0-9-]+)?$/;
+  allClustering: any = [];
+  selectedLabels: string[] = [];
+  selectedHosts: string[] = [];
+  action: any;
+  resource: string;
+  icons = Icons;
+  domainSettingObject: any;
+  modalData$ = this.smbService.modalData$;
+
+  constructor(
+    private hostService: HostService,
+    private formBuilder: CdFormBuilder,
+    private cd: ChangeDetectorRef,
+    public smbService: SmbService,
+    public actionLabels: ActionLabelsI18n,
+    private orchService: OrchestratorService,
+    private modalService: ModalCdsService,
+    private taskWrapperService: TaskWrapperService,
+    private router: Router
+  ) {
+    super();
+    this.resource = $localize`Cluster`;
+  }
+  ngOnInit() {
+    this.action = this.actionLabels.CREATE;
+    this.smbService.modalData$.subscribe((data) => {
+      this.domainSettingObject = data;
+      this.smbForm.get('domain_setting').setValue(data?.realm);
+    });
+    this.createForm();
+
+    this.action = this.actionLabels.CREATE;
+    this.hostsAndLabels$ = forkJoin({
+      hosts: this.hostService.getAllHosts(),
+      labels: this.hostService.getLabels()
+    }).pipe(
+      map(({ hosts, labels }) => ({
+        hosts: hosts.map((host: any) => ({ content: host['hostname'] })),
+        labels: labels.map((label: string) => ({ content: label }))
+      }))
+    );
+    this.orchStatus$ = this.orchService.status();
+    this.allClustering = ['Default', 'Always', 'Never'];
+  }
+
+  createForm() {
+    this.smbForm = this.formBuilder.group({
+      cluster_id: new FormControl('', {
+        validators: [Validators.required]
+      }),
+      auth_mode: [
+        '',
+        {
+          validators: [Validators.required]
+        }
+      ],
+      intent: [
+        '',
+        {
+          validators: [Validators.required]
+        }
+      ],
+      domain_setting: [''],
+      user_group_setting: [[]],
+      placement: [{}],
+      hosts: [[]],
+      label: [
+        null,
+        [
+          CdValidators.requiredIf({
+            placement: 'label'
+          })
+        ]
+      ],
+      custom_dns: new FormArray([]),
+      jointSources: new FormArray([]),
```

Comment:
It should be, `joinSources`, right?
```suggestion
      joinSources: new FormArray([]),
```